File: projects/native-apps/ios/App/LaunchIntroModel.swift
import Foundation

@MainActor
final class LaunchIntroModel: ObservableObject {
	@Published private(set) var phase: LaunchIntroPhase
	@Published private(set) var isMuted: Bool
	private let music: IntroMusic
	private let mutePreference: IntroMutePreference

	init(
		seen: LaunchIntroSeen,
		music: IntroMusic,
		mutePreference: IntroMutePreference,
		reduceMotion: Bool,
		isLoggedIn: Bool
	) {
		self.music = music
		self.mutePreference = mutePreference
		self.isMuted = mutePreference.isMuted
		self.phase = LaunchIntro.initialPhase(isFirstLaunch: seen.claim(), reduceMotion: reduceMotion)
		music.setMuted(isMuted)
		if LaunchIntro.playsMusic(isLoggedIn: isLoggedIn, isForeground: true) {
			music.start()
		}
	}

	var overlay: LaunchIntroOverlay {
		LaunchIntro.overlay(for: phase)
	}

	func end(_ reason: LaunchIntroEnd) {
		phase = LaunchIntro.next(after: phase, end: reason)
		if reason == .skipped {
			music.seek(LaunchIntro.videoDuration)
		}
	}

	func fadeCompleted() {
		guard phase == .fading else { return }
		phase = .finished
	}

	func replay() {
		phase = .playing
		isMuted = false
		mutePreference.setMuted(false)
		music.setMuted(false)
		music.restart()
	}

	func toggleMute() {
		isMuted.toggle()
		mutePreference.setMuted(isMuted)
		music.setMuted(isMuted)
	}

	func sync(isLoggedIn: Bool, isForeground: Bool) {
		if LaunchIntro.playsMusic(isLoggedIn: isLoggedIn, isForeground: isForeground) {
			music.start()
		} else {
			music.stop()
		}
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/app/LaunchIntroModel.kt
package com.readplace.android.app

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

interface IntroMusic {
	fun start()

	fun stop()

	fun restart()

	fun seek(seconds: Double)

	fun setMuted(muted: Boolean)
}

class LaunchIntroModel(
	seen: LaunchIntroSeen,
	private val music: IntroMusic,
	private val mutePreference: IntroMutePreference,
	reduceMotion: Boolean,
	isLoggedIn: Boolean,
) {
	private val _isMuted = MutableStateFlow(mutePreference.isMuted)
	val isMuted: StateFlow<Boolean> = _isMuted.asStateFlow()

	private val _phase = MutableStateFlow(
		LaunchIntro.initialPhase(isFirstLaunch = seen.claim(), reduceMotion = reduceMotion),
	)
	val phase: StateFlow<LaunchIntroPhase> = _phase.asStateFlow()

	init {
		music.setMuted(_isMuted.value)
		if (LaunchIntro.playsMusic(isLoggedIn = isLoggedIn, isForeground = true)) {
			music.start()
		}
	}

	val overlay: LaunchIntroOverlay
		get() = LaunchIntro.overlay(_phase.value)

	fun end(reason: LaunchIntroEnd) {
		_phase.value = LaunchIntro.next(after = _phase.value, end = reason)
		if (reason == LaunchIntroEnd.SKIPPED) {
			music.seek(LaunchIntro.VIDEO_DURATION)
		}
	}

	fun fadeCompleted() {
		if (_phase.value != LaunchIntroPhase.FADING) return
		_phase.value = LaunchIntroPhase.FINISHED
	}

	fun replay() {
		_phase.value = LaunchIntroPhase.PLAYING
		_isMuted.value = false
		mutePreference.isMuted = false
		music.setMuted(false)
		music.restart()
	}

	fun toggleMute() {
		val muted = !_isMuted.value
		_isMuted.value = muted
		mutePreference.isMuted = muted
		music.setMuted(muted)
	}

	fun sync(isLoggedIn: Boolean, isForeground: Boolean) {
		if (LaunchIntro.playsMusic(isLoggedIn = isLoggedIn, isForeground = isForeground)) {
			music.start()
		} else {
			music.stop()
		}
	}
}
